pkg/controller/nstemplatetier/nstemplatetier_controller.go

```diff
@@ -0,0 +1,251 @@
+		return reconcile.Result{}, err
+	}
+
+	activeTemplateUpdateRequests, err := r.activeTemplateUpdateRequests(instance)
+	if err != nil {
+		logger.Error(err, "unable to get active TemplateUpdateRequests")
+		return reconcile.Result{}, err
+	}
+
+	if activeTemplateUpdateRequests < MaxPoolSize {
+		members := r.retrieveMemberClusters()
+	loop:
+		for _, member := range members {
+			// create a TemplateUpdateRequest if active count < max pol size
+			// find a MasterUserRecord which is not already up-to-date
+			// and for which there is no TemplateUpdateRequest yet
+			// fetch by subsets of "MaxPoolSize + 1" size until a MasterUserRecord candidate is found
+			murs := toolchainv1alpha1.MasterUserRecordList{}
+			matchingLabels, err := newLabelSelector(instance, *member)
+			if err != nil {
+				logger.Error(err, "unable to get MasterUserRecords to update")
+				return reconcile.Result{}, err
+			}
+			r.client.List(context.Background(), &murs,
+				client.InNamespace(instance.Namespace),
+				client.Limit(MaxPoolSize+1),
+				matchingLabels,
+			)
+			logger.Info("listed MasterUserRecords", "count", len(murs.Items), "cluster", member.Name)
+			for _, mur := range murs.Items {
+				for _, account := range mur.Spec.UserAccounts {
+					accountTmpls := account.Spec.NSTemplateSet
+					logger.Info("checking mur account", "name", mur.Name, "tier", accountTmpls.TierName)
+					if accountTmpls.TierName != instance.Name {
+						// skip if the user account is not based on the current Tier
+						continue
+					}
+					// compare all TemplateRefs in the current MUR.UserAccount vs NSTemplateTier instance
+					if !reflect.DeepEqual(templateRefs(accountTmpls), templateRefs(instance.Spec)) {
+						logger.Info("creating a TemplateUpdateRequest to update the MasterUserRecord", "name", mur.Name, "tier", instance.Name)
+						r.client.Create(context.TODO(), &toolchainv1alpha1.TemplateUpdateRequest{
+							ObjectMeta: metav1.ObjectMeta{
+								Name: mur.Name,
+								Labels: map[string]string{
+									toolchainv1alpha1.NSTemplateTierNameLabelKey: instance.Name,
+								},
+							},
+							Spec: toolchainv1alpha1.TemplateUpdateRequestSpec{
+								TierName:         instance.Name,
+								Namespaces:       instance.Spec.Namespaces,
+								ClusterResources: instance.Spec.ClusterResources,
+							},
+						})
+						break loop // create a single TemplateUpdateRequest resource per reconcile loop
+					}
+				}
+			}
+		}
+	}
+	logger.Info("done with creating TemplateUpdateRequest resources after update of NSTemplateTier", "tier", instance.Name)
+	return reconcile.Result{}, nil
+}
+
+// newLabelSelector creates a label selector to find MasterUserRecords which are not up-to-date with
+// the templateRefs of the given NSTemplateTier.
+//
+// (longer explanation)
+// newLabelSelector creates a selector to find MasterUserRecords with the following labels:
+// - toolchain.dev.openshift.com/<member>-templates-tier: <tier>
```

Comment:
Why do we care about `<member>` here? Shouldn't we just use `toolchain.dev.openshift.com/templates-tier` and `toolchain.dev.openshift.com/templates-tier-hash` labels instead?
I'm a bit concerned about binding MURs to target clusters. Our UserAccounts are bind to target/member clusters. Not MURs. And MURs can have multiple UserAccounts with different target clusters. We don't use it now but we will when supporting namespace sharing between clusters and/or migrating between clusters.

Reply:
hmm, I thought a user could have different tier, depending on the cluster she uses. At least the design of the MasterUserRecord allows it, AFAICS.

Reply:
I see the reasoning.. But there are some issues with this approach:
- You have to iterate over all member clusters to fetch all MUR to update. So, if there are 10 member clusters you will have to list murs 10 times for each tier.
- The name of the label can't be longer than 63 characters. If we have too long cluster name then we may exceed that limit. As the matter of fact it won't work with our current crt-stage clusters. For our member cluster the label would be `toolchain.dev.openshift.com/member-member-8bs24-templates-tier-hash` which is 68 characters.

What if we use `toolchain.dev.openshift.com/<tier-name>-tier-version` label instead of these two labels?
So, we can have multiple labels (one per tier) within a MUR:
- `toolchain.dev.openshift.com/advanced-tier-version: xxx`
- `toolchain.dev.openshift.com/basic-collaborator-tier-version: yyy`

And we can look up all MURs with the outdated advanced tier by using `toolchain.dev.openshift.com/advanced-tier-version != zzz` selector.

Also I was thinking what is better to use here... `-version` or `-hash` as the label name. We could use `-version-hash` but it's too long... WDYT? I'm OK with both options.

Reply:
> I see the reasoning.. But there are some issues with this approach:
> 
> * You have to iterate over all member clusters to fetch all MUR to update. So, if there are 10 member clusters you will have to list murs 10 times for each tier.

That's right :/ 

> * The name of the label can't be longer than 63 characters. If we have too long cluster name then we may exceed that limit. As the matter of fact it won't work with our current crt-stage clusters. For our member cluster the label would be `toolchain.dev.openshift.com/member-member-8bs24-templates-tier-hash` which is 68 characters.

Damn'it! (as a side note, why did we name the cluster `member-member-8bs24`?)

> 
> What if we use `toolchain.dev.openshift.com/<tier-name>-tier-version` label instead of these two labels?
> So, we can have multiple labels (one per tier) within a MUR:
> 
> * `toolchain.dev.openshift.com/advanced-tier-version: xxx`
> * `toolchain.dev.openshift.com/basic-collaborator-tier-version: yyy`
> 
> And we can look up all MURs with the outdated advanced tier by using `toolchain.dev.openshift.com/advanced-tier-version != zzz` selector.
>

ok, I can do that! BTW, I realize that the current solution in the PR would not work because we use the TemplateUpdateRequests have the same name as their associated MUR, so we would have an error if a user had accounts with the same tier to update on several clusters! 🤦‍♂️ 
 
> Also I was thinking what is better to use here... `-version` or `-hash` as the label name. We could use `-version-hash` but it's too long... WDYT? I'm OK with both options.

I'd rather stay with `-hash` to keep the label slightly shorter ;)

Reply:
actually, the selector should check that the MUR has a label with key `toolchain.dev.openshift.com/<tiername>-tier-hash` but with value `!= <hash>`, because we don't want to select MUR associated with another tier (I will have to test that too, BTW!)

Reply:
also, we need to have this new `toolchain.dev.openshift.com/<tiername>-tier-hash` label set on all MasterUserRecords prior to be able to update the templates. Otherwise, none of the MasterUserRecords will be selected!

Reply:
done in 6a6ff2a

Reply:
> > actually, the selector should check that the MUR has a label with key toolchain.dev.openshift.com/-tier-hash but with value `!= <hash>`
> 
> This is exactly what I said, no? :)

well, I wanted to emphasize that with a single selector such as:
```
toolchain.dev.openshift.com/advanced-tier-version != zzz
```
we would select MURs using an outdated "advanced" tier, but we may also select all MURs NOT using this tier. Hence, we need to verify that the `toolchain.dev.openshift.com/advanced-tier-version` label exists, whatever its value.